firmware/robot2015/src-kckr/main.c

```diff
@@ -1,13 +1,237 @@
 #include <avr/io.h>
+#include <avr/interrupt.h>
 #include <util/delay.h>
+// TODO: Make this path less explicit!
+#include "kicker_commands.h"
+#include "pins.h"
 
-int main(void) {
-    DDRA = 0xFF;
+#define NO_COMMAND 0
+#define IGNORE_CS 0
 
+#define TIMING_CONSTANT 125
+#define VOLTAGE_READ_DELAY_MS 100
+
+// State of the ATTINY kicking/chipping
+typedef enum { OFF, ON, ACTING } state_type;
+volatile state_type state_ = ON;
+
+// Used to time kicks and chips
+volatile unsigned millis_left_ = 0;
+
+// Used to keep track of current button state
+volatile int kick_db_held_down_ = 0;
+volatile int chip_db_down_ = 0;
+volatile int charge_db_down_ = 0;
+
+uint8_t cur_command_ = NO_COMMAND;
+
+// always up-to-date voltage so we don't have to get_voltage() inside interupts
+uint8_t last_voltage_ = 0;
+// function that does a voltage reading
+uint8_t get_voltage();
+// executes a command coming from SPI
+void execute_cmd(uint8_t, uint8_t);
+
+void main() {
+    /* Port direction - setting outputs */
+    DDRA |= _BV(KICK_PIN) | _BV(CHIP_PIN) |
+            _BV(CHARGE_PIN);  // MISO is handled by CS interrupt
+    // ensure N_KICK_CS is input
+    DDRA &= ~_BV(N_KICK_CS_PIN);
+    // ensure debug buttons are inputs
+    DDRB &= ~_BV(DB_KICK_PIN) & ~_BV(DB_CHIP_PIN) & ~_BV(DB_CHG_PIN);
+    // Which DDRB = 0 and PORTB = 1, these are configured as pull-up inputs
+    PORTB |= _BV(DB_KICK_PIN) | _BV(DB_CHIP_PIN) | _BV(DB_CHG_PIN);
+    // Chip Select Interrupt
+    if (IGNORE_CS)
+        // Because we aren't using CS here, MISO must be manually set as an
+        // output instead of being managed by the interrupt
+        DDRA |= _BV(MISO_PIN);
+    else
+        GIMSK |= _BV(PCIE0);  // Enable interrupts for PCINT0-PCINT7
+    // Debug Button Interrupts
+    GIMSK |= _BV(PCIE1);           // Enable interrupts for PCINT8-PCINT11
+    PCMSK0 |= _BV(INT_N_KICK_CS);  // Enable on N_KICK_CS
+    // Enable interrupts on debug buttons
+    PCMSK1 |= _BV(INT_DB_KICK) | _BV(INT_DB_CHIP) | _BV(INT_DB_CHG);
+    // SPI init - Pg. 120
+    USICR |= _BV(USIWM0)     // 3 Wire Mode MISO, DI, USCK - Pg. 124
+             | _BV(USICS1)   // External, negative edge clock - Pg. 125
```

Comment:
Should this say positive edge?  It looks like that's what it's being configured for according to the datasheet.